mathics/core/expression.py

```diff
@@ -186,9 +210,25 @@ def __init__(self, head, *elements, **kwargs):
 
         # Set some properties on elements that help us speed up evaluation.
         # These are set in self._build_elements(elements)
-        #    self._elements_fully_evaluated, self._is_flat, self._is_sorted
-
-        self.elements = elements
+        #    self._elements_fully_evaluated, self._is_flat, self._is_ordered
+
+        conversion_fn = kwargs.pop("element_conversion_fn", from_python)
+        element_properties = kwargs.pop("element_properties", None)
+
+        # Note: We don't allow specifying "_is_ordered"
+        # when "_elements_fully_evaluated" is False. And I suppose this
+        # might be right since we can't really know ordering if we things
+        # are not fully evaluated.
+        if element_properties is not None:
+            for field in ("_elements_fully_evaluated", "_is_flat", "_is_ordered"):
+                setattr(self, field, element_properties.get(field, False))
+            self._elements = (
+                elements
+                if self._elements_fully_evaluated
```

Comment:
I am missing the point here.  The code does the right thing for these cases.

Reply:
Yes, but now, what would be the right parameters to pass  in element_properties?

Reply:
For `Expression("Plus", 50*[1])` doing it this way works and sets the properties correctly. 

As done in the `test_expression_constructor` you can also use
`Expression("Plus"  50*[1], element_properties={"_elements_fully_evaluated": True, "_is_flat": True, "is_ordered: True})` if you want be explicit about the elements properties so Expression does not try to evaluate, flatten or order the elements.

As for `Expression("Plus", Symbol("Global`a"), Integer(1))` best to just leave it as is. If you wanted to add element properties it would be what you said: `element_properties={"_elements_fully_evaluated: False", "_is_flat": True}`. 

You could specify that this isn't ordered, but that is also implied when elements are not fully evaluated. (How can something  be ordered when you don't have a value for everything).

This kind of thing isn't causing a problem. 

The problem place I am concerned about is when we have long lists. You find this for example in plotting.

Reply:
> For `Expression("Plus", 50*[1])` doing it this way works and sets the properties correctly.
> 
> As done in the `test_expression_constructor` you can also use `Expression("Plus" 50*[1], conversion_fn=Integer, element_properties={"_elements_fully_evaluated": True, "_is_flat": True, "is_ordered: True})` if you want be explicit about the elements properties so Expression does not try to evaluate, flatten or order the elements.
>

Following the line 225,
```
self._elements = (
                elements
                if self._elements_fully_evaluated 
                else (conversion_fn(element) for element in elements)
            )
```
if you set that the element is fully evaluated, then the conversion shouldn´t be done. 
So, I do not see where
```Expression("Plus", 50*[1], conversion_fn=Integer, element_properties={"_elements_fully_evaluated": True, "_is_flat": True, "is_ordered: True})```
applies conversion_fn over the elements. Probably it happens in another place.

 
> As for `` Expression("Plus", Symbol("Global`a"), Integer(1)) `` best to just leave it as is. If you wanted to add element properties it would be what you said: `element_properties={"_elements_fully_evaluated: False", "_is_flat": True}`.
> 
> You could specify that this isn't ordered, but that is also implied when elements are not fully evaluated. (How can something be ordered when you don't have a value for everything?)
> 

Agree

> This kind of thing isn't causing a problem.
> 
> The place I was concerned about is when we have long lists or lots of data. You find this for example in plotting.

OK, but if you want to build a Plot, usually you do not have a mixture of Python and Mathics native numbers: well you pass a NumPy array, well you pass a ``List`` ``Expression`` where all the elements are native numbers.
In any case, all this conversion stuff comes up just to make some typical code easier to read and write. 
It is unlike, (and probably, we should discourage)  to have code like

```
a=sympy.Symbol("a")
b = 20*[1.]
c = Icecream("mint")
expr = Expression("F", a , b, c, conversion_fn=funny_converter,  {"_elements_fully_evaluated: False", "_is_flat": True})
```

Also, I do not see the point of allowing
```
data = 50*[1]
expr = Expression(SymbolList, *data, conversion_fn=Integer)
```
which can be written in a clearer (and sometimes more efficient) way as
```
data = 50*[Integer1]
expr = Expression(SymbolList, *data, conversion_fn=None)
```
(or even)
```
data = 50*[Integer1]
expr = List(*data, conversion_fn=None)  # not sure there if this is supported now
```

Reply:
Here is the example:
```
>> from mathics.session import MathicsSession
>> from mathics.core.expression import Expression, Integer
>> expr = Expression("Plus", 10*[1], element_conversion_fn=Integer, element_properties={"_elements_fully_evaluated": True,"_is_flat": True,"_is_ordered": True})
>> type(expr.elements[0])
<class 'list'>
```

Reply:
> > For `Expression("Plus", 50*[1])` doing it this way works and sets the properties correctly.
> > As done in the `test_expression_constructor` you can also use `Expression("Plus" 50*[1], conversion_fn=Integer, element_properties={"_elements_fully_evaluated": True, "_is_flat": True, "is_ordered: True})` if you want be explicit about the elements properties so Expression does not try to evaluate, flatten or order the elements.
> 
> Following the line 225,
> 
> ```
> self._elements = (
>                 elements
>                 if self._elements_fully_evaluated 
>                 else (conversion_fn(element) for element in elements)
>             )
> ```
> 
> if you set that the element is fully evaluated, then the conversion shouldn´t be done. So, I do not see where `Expression("Plus", 50*[1], conversion_fn=Integer, element_properties={"_elements_fully_evaluated": True, "_is_flat": True, "is_ordered: True})` applies conversion_fn over the elements. Probably it happens in another place.


My mistake here.  If you are passing a conversion function, you _can't_ set element properties as the code is currently written. The idea here is that if you have a conversion function `Expression._build_elements()` does the computation; it not has to iterate over the list anyway, and the property calculations are simple and not expensive.

So that should have been more simply: `Expression("Plus", 50*[1], conversion_fn=Integer)` . 

You have convinced me that this interface is a little too complicated. But right now this is not where I want to focus on. 

My take right now is to keep https://github.com/Mathics3/mathics-core/pull/282#partial-pull-merging off to the side for now and come back to it after we've made a full pass over evaluation. 

... 
>>  The place I was concerned about is when we have long lists or lots of data. You find this for example in plotting.
> 
> OK, but if you want to build a Plot, usually you do not have a mixture of Python and Mathics native numbers: well you pass a NumPy array, well you pass a `List` `Expression` where all the elements are native numbers. In any case, all this conversion stuff comes up just to make some typical code easier to read and write. It is unlike, (and probably, we should discourage) to have code like
> 
> ```
> a=sympy.Symbol("a")
> b = 20*[1.]
> c = Icecream("mint")
> expr = Expression("F", a , b, c, conversion_fn=funny_converter,  {"_elements_fully_evaluated: False", "_is_flat": True})
> ```

Right now I am thinking that there might be a special SympyExpression class for wrapping SymPy expressions that are used in Mathics. `from_sympy()`  would be used to extract a Mathics expression from this, but it likely that things would be kept in SymPy format as long a possible. This might have come up in https://github.com/Mathics3/mathics-core/issues/20#issuecomment-930319860 and the may get around slowness in https://github.com/Mathics3/mathics-core/issues/103#issuecomment-1009311644

I am not certain about this, but it is a possibility. 

> 
> Also, I do not see the point of allowing
> 
> ```
> data = 50*[1]
> expr = Expression(SymbolList, *data, conversion_fn=Integer)
> ```
> 
> which can be written in a clearer (and sometimes more efficient) way as
> 
> ```
> data = 50*[Integer1]
> expr = Expression(SymbolList, *data, conversion_fn=None)
> ```
> 
> (or even)
> 
> ```
> data = 50*[Integer1]
> expr = List(*data, conversion_fn=None)  # not sure there if this is supported now
> ```

I either don't understand what you are talking about and possibly don't agree, but I don't think it worth the effort debate any of this.

Reply:
I wrote: 
> You have convinced me that this interface is a little too complicated. 

However in looking at use cases such as those found in Range of constructing.py, I think this may be the right thing. 

It may turn out that we don't need to set element properties at all (or rarely) when we have a pre-formed Expression.

I am not certain about that. But this is something that can be dealt with later.